feat(config): `platformOptions` (#30582)
renovatebot/renovate: docs/usage/self-hosted-configuration.md

## platform

## platformVersion

> Specify this string for Renovate to skip API checks and provide GitLab/Bitbucket server version directly.
> Particularly useful with GitLab's `CI_JOB_TOKEN` to authenticate Renovate or to reduce API calls.

Reviewer: ```suggestion
When you set a `platformVersion` string, Renovate skips a API call to GitLab/Bitbucket.
Instead of fetching the version from the API, Renovate uses the version in the `platformVersion` string.

Setting a `platformVersion` manually is handy when you:

- use GitLab's `CI_JOB_TOKEN` to authenticate Renovate, or
- want to reduce the number of API calls
```

> Read [platform details](modules/platform/gitlab/index.md) to learn why we need the server version on GitLab.

Reviewer: ```suggestion
Read [platform details](modules/platform/gitlab/index.md) to learn why Renovate must know the GitLab server version.
```

<!-- prettier-ignore -->
!!! note

>     Only valid within [platformOptions](configuration-options.md#platformoptions) object.

Reviewer: ```suggestion
    You must put the `platformVersion` config option in the [`platformOptions`](configuration-options.md#platformoptions) object.
    If you put the `platformVersion` anywhere else, it will not work.
```

## platform

> ## platformVersion

Author: should we add `platformOptions` in both `configration-options.md` & `self-hosted-configuration.md`?

> ## platformVersion

Reviewer 2: yes, I think we need to.

> ## platformVersion

Author: This will disturb the docs test, as we do not expect same option to be present in both `configuration-options.md` & `self-hosted-configuration.md`.

Should I add an exception in the tests, or something else?

> ## platformVersion

Reviewer: > Should I add an exception in the tests, or something else?

I don't like exceptions to tests. That makes it harder to understand what and why we're testing something.

> This will disturb the docs test, as we do not expect same option to be present in both `configuration-options.md` & `self-hosted-configuration.md`.

The test failure also shows that using the same name for a config option in two places is confusing.

Questions:

- Does `platformVersion` behave _exactly_ the same on Mend-hosted app and self-hosted?
- If there is a difference in behavior, should we use different names for each `platformVersion`? For example: `selfhostedPlatformVersion` and a plain `platformVersion`?

## platform

## platformOptions

Use `platformOptions` to set platform-specific configuration options provided by Renovate.

Example:

```json
{
  "platformOptions": {

>     "gitLabIgnoreApprovals": true

Reviewer 2: ```suggestion
    "platformVersion": "1.2.3"
```

## platform

> ## platformOptions

Reviewer 3: I'm confused by why this is in both self-hosted and repo docs. Why is that?

Author: While `platformOptions` can appear in both configs, the specific child options within it are constrained depending on whether they are part of the admin (self-hosted) or repo config.

For example:

``` platformOptions: { gitlabIgnoreApprovals: true } ``` can be used in both the admin (self-hosted) and repository configurations.
But, ``` platformOptions: { platformVersion: '1.2.3' } ``` can only be used in the admin (self-hosted) configuration because `platformVersion` is specific to admin config.

To make this distinction clearer, I've added documentation for `platformOptions` in both the self-hosted and repository docs. In each, I've listed which children are allowed in the respective configs (admin or repo).

Reviewer 3: I don't like this approach - too much chance of confusion or mistake.

Reviewer 2: @rarkins Any other suggestion how to handle platform options? Maybe naming the other `repoOptions` for repo configs?

Reviewer 3: What was the idea behind creating this combined platformOptions option? to which functions are they passed? e.g. for initPlatform() only the global ones are passed, but for ensurePr() both global and repo are passed?

Author: The idea was to organise existing platform options under one parent to make it easier for users to find them in the docs.
Global ones are passed to `initPlatform` and repo ones are passed to `ensurePr`. 

The global ones are filtered out in the `getPlatformPrOptions` fn before being passed to `ensurePr`.

I agree, though that this is confusing.

Reviewer 3: Could it just be `prOptions` for short? `platformOptions` are passed to `initPlatform()` and `prOptions` are passed to `ensurePr()`

Reviewer 2: sounds good 👍
